cmd/contact-auditor/main.go

```diff
@@ -0,0 +1,203 @@
+package main
+
+import (
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	"github.com/jmhodges/clock"
+	"github.com/letsencrypt/boulder/cmd"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/policy"
+)
+
+var queryInterrupted bool
+
+type contactAuditor struct {
+	db     *sql.DB
+	logger blog.Logger
+	clk    clock.Clock
+	grace  time.Duration
+}
+
+type result struct {
+	// Receiver for the `id` column.
+	ID int64
+
+	// Receiver for the `contact` column.
+	Contact  []byte
+	contacts []string
+}
+
+func (r *result) unmarshalContact() error {
+	var contact []string
+	err := json.Unmarshal(r.Contact, &contact)
+	if err != nil {
+		return err
+	}
+	r.contacts = append(r.contacts, contact...)
+	return nil
+}
+
+func (r *result) validateContact() error {
+	// Setup a buffer to store any validation problems we encounter.
+	var probsBuff strings.Builder
+
+	// Helper to write validation problems to our buffer.
+	writeProb := func(contact string, prob string) {
+		// Add validation problem to buffer.
+		probsBuff.WriteString(fmt.Sprintf("[ %q: %q ] ", contact, prob))
+	}
+
+	for _, contact := range r.contacts {
+		if strings.HasPrefix(contact, "mailto:") {
+			err := policy.ValidEmail(strings.TrimPrefix(contact, "mailto:"))
+			if err != nil {
+				writeProb(contact, err.Error())
+			}
+		} else {
+			writeProb(contact, "missing 'mailto:' prefix")
+		}
+	}
+
+	if probsBuff.String() != "" {
+		return errors.New(probsBuff.String())
+	}
+	return nil
+}
+
+// beginAuditQuery executes the audit query and returns a cursor used to
+// stream the results.
+func (c contactAuditor) beginAuditQuery() (*sql.Rows, error) {
+	rows, err := c.db.Query(
+		fmt.Sprintf(
+			"%s '%s';",
+			`SELECT DISTINCT r.id, r.contact
+			FROM registrations AS r
+				INNER JOIN certificates AS c on c.registrationID = r.id
+			WHERE r.contact NOT IN ('[]', 'null')
+				AND c.expires >=`,
+			c.clk.Now().Add(-c.grace).Format("2006-01-02T15:04:05Z")),
```

Comment:
Why do you interpolate both the query and the timestamp into a string, instead of interpolating the timestamp directly into the query?

Reply:
😆 because I'm a dork who didn't realize that you could use the formatting verbs in a string literal. Lesson learned!